File: app/src/lib/game/deck.ts
import { SHAPES, SUIT_VALUES, WHOT_VALUE } from './types.js';
import type { Card, SuitCard, WhotCard } from './types.js';

const WHOT_COUNT = 5;

export function createDeck(): Card[] {
	const suitCards: SuitCard[] = SHAPES.flatMap((shape) =>
		SUIT_VALUES.map((value): SuitCard => ({ kind: 'suit', shape, value }))
	);

	const whotCards: WhotCard[] = Array.from(
		{ length: WHOT_COUNT },
		(): WhotCard => ({ kind: 'whot', value: WHOT_VALUE })
	);

	return [...suitCards, ...whotCards];
}

// Fisher-Yates — O(n), every permutation equally likely
export function shuffle<T>(items: readonly T[]): T[] {
	const result = [...items];
	for (let i = result.length - 1; i > 0; i--) {
		const j = Math.floor(Math.random() * (i + 1));
		const temp = result[i];
		result[i] = result[j];
		result[j] = temp;
	}
	return result;
}

export function createShuffledDeck(): Card[] {
	return shuffle(createDeck());
}

File: server/src/game/deck.rs
use crate::game::types::{Card, SHAPES, SUIT_VALUES, WHOT_COUNT};
use rand::seq::SliceRandom;

pub fn create_deck() -> Vec<Card> {
    let mut deck: Vec<Card> = SHAPES
        .iter()
        .flat_map(|&shape| {
            SUIT_VALUES
                .iter()
                .map(move |&value| Card::Suit { shape, value })
        })
        .collect();
    for _ in 0..WHOT_COUNT {
        deck.push(Card::Whot);
    }
    deck
}

pub fn shuffled_deck(rng: &mut impl rand::Rng) -> Vec<Card> {
    let mut deck = create_deck();
    deck.shuffle(rng);
    deck
}
